connection_pool/connection_pool.go

```diff
@@ -0,0 +1,712 @@
+func (connPool *ConnectionPool) InsertAsync(space interface{}, tuple interface{}) *tarantool.Future {
+	conn, err := connPool.getConnByMode(RW, nil)
+	if err != nil {
+		return nil
+	}
+
+	return conn.InsertAsync(space, tuple)
+}
+
+// ReplaceAsync sends "insert or replace" action to tarantool and returns Future.
+// If tuple with same primary key exists, it will be replaced.
+func (connPool *ConnectionPool) ReplaceAsync(space interface{}, tuple interface{}) *tarantool.Future {
+	conn, err := connPool.getConnByMode(RW, nil)
+	if err != nil {
+		return nil
+	}
+
+	return conn.ReplaceAsync(space, tuple)
+}
+
+// DeleteAsync sends deletion action to tarantool and returns Future.
+// Future's result will contain array with deleted tuple.
+func (connPool *ConnectionPool) DeleteAsync(space, index interface{}, key interface{}) *tarantool.Future {
+	conn, err := connPool.getConnByMode(RW, nil)
+	if err != nil {
+		return nil
+	}
+
+	return conn.DeleteAsync(space, index, key)
+}
+
+// UpdateAsync sends deletion of a tuple by key and returns Future.
+// Future's result will contain array with updated tuple.
+func (connPool *ConnectionPool) UpdateAsync(space, index interface{}, key, ops interface{}) *tarantool.Future {
+	conn, err := connPool.getConnByMode(RW, nil)
+	if err != nil {
+		return nil
+	}
+
+	return conn.UpdateAsync(space, index, key, ops)
+}
+
+// UpsertAsync sends "update or insert" action to tarantool and returns Future.
+// Future's sesult will not contain any tuple.
+func (connPool *ConnectionPool) UpsertAsync(space interface{}, tuple interface{}, ops interface{}) *tarantool.Future {
+	conn, err := connPool.getConnByMode(RW, nil)
+	if err != nil {
+		return nil
+	}
+
+	return conn.UpsertAsync(space, tuple, ops)
+}
+
+// CallAsync sends a call to registered tarantool function and returns Future.
+// It uses request code for tarantool 1.6, so future's result is always array of arrays
+func (connPool *ConnectionPool) CallAsync(functionName string, args interface{}, userMode ...Mode) *tarantool.Future {
+	conn, err := connPool.getConnByMode(RW, userMode)
+	if err != nil {
+		return nil
+	}
+
+	return conn.CallAsync(functionName, args)
+}
+
+// Call17Async sends a call to registered tarantool function and returns Future.
+// It uses request code for tarantool 1.7, so future's result will not be converted
+// (though, keep in mind, result is always array)
+func (connPool *ConnectionPool) Call17Async(functionName string, args interface{}, userMode ...Mode) *tarantool.Future {
+	conn, err := connPool.getConnByMode(RW, userMode)
+	if err != nil {
+		return nil
+	}
+
+	return conn.Call17Async(functionName, args)
+}
+
+// EvalAsync sends a lua expression for evaluation and returns Future.
+func (connPool *ConnectionPool) EvalAsync(expr string, args interface{}, userMode ...Mode) *tarantool.Future {
+	conn, err := connPool.getConnByMode(RW, userMode)
+	if err != nil {
+		return nil
```

Comment:
If we use `return future.fail(conn, err)` we need `conn`, but there could be no connection according to mode and we need to return an error about it, so I decided to return an error as second parameter:

```golang
func (connPool *ConnectionPool) InsertAsync(space interface{}, tuple interface{}) (*tarantool.Future, error) {
	conn, err := connPool.getConnByMode(RW, nil)
	if err != nil || conn == nil {
		return nil, err
	}

	return conn.InsertAsync(space, tuple), nil
}
```

Reply:
I agree that it would be great if we implement it consistently, but right now `newFuture` method is handled by `Connection` object:

```golang
func (conn *Connection) newFuture(requestCode int32) (fut *Future)
```

I suppose we should have some new methods to create future without connection.

Reply:
Done.

Reply:
Awesome!